### PR TITLE
executor: support Chunk for SelectionExec

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

File: util/chunk/chunk.go [hunk 1]
```diff
@@ -347,6 +347,11 @@ type Row struct {
 	idx int
 }
 
+// ID returns the row index of Chunk.
+func (r Row) ID() int {
```
Comment (coocood) [inline util/chunk/chunk.go]: I think we should use `Idx` for consistency.

**Comment by coocood** [comment]
@zz-jason 
Should consider vectorizable?

File: expression/chunk_executor.go [hunk 2]
```diff
@@ -234,3 +234,47 @@ func executeToString(sc *stmtctx.StatementContext, expr Expression, fieldType *t
...
+			}
+			selected[row.Idx()] = selected[row.Idx()] && !isNull && (filterResult != 0)
+		}
+	}
+	return selected, nil
+}
+
+// UnVectorizedFilter applys a list of filters to a Chunk and returns a bool slice, which indicates whether a row is passed the filters.
+// Filters is not executed vectorized.
+func UnVectorizedFilter(ctx context.Context, filters []Expression, input *chunk.Chunk, selected []bool) ([]bool, error) {
```
Comment (coocood) [inline expression/chunk_executor.go]: Our normal tests do not cover this, we should add a specific test case for it.

**Comment by coocood** [comment]
LGTM

File: executor/executor.go [hunk 3]
```diff
@@ -571,6 +594,51 @@ func (e *SelectionExec) Next(goCtx goctx.Context) (Row, error) {
 	}
 }
 
+// NextChunk implements the Executor NextChunk interface.
+func (e *SelectionExec) NextChunk(chk *chunk.Chunk) error {
+	chk.Reset()
+	for {
+		for ; e.inputRow != e.childrenResults[0].End(); e.inputRow = e.inputRow.Next() {
+			if e.batched {
```
Comment (tiancaiamao) [inline executor/executor.go]: I think you can split batch  and non-batch logic into two different function.
Keep the function small and compact make the code more readable.

**Comment by zz-jason** [comment]
@tiancaiamao PTAL again

File: executor/executor.go [hunk 4]
```diff
@@ -571,6 +594,67 @@ func (e *SelectionExec) Next(goCtx goctx.Context) (Row, error) {
...
+		return errors.Trace(e.unBatchedNextChunk(chk))
+	}
+
+	for {
+		for ; e.inputRow != e.childrenResults[0].End(); e.inputRow = e.inputRow.Next() {
+			if !e.selected[e.inputRow.Idx()] {
+				continue
+			}
+			chk.AppendRow(0, e.inputRow)
+			if chk.NumRows() == e.ctx.GetSessionVars().MaxChunkSize {
```
Comment (XuHuaiyu) [inline executor/executor.go]: 1. move this check to line 610 thus line 612 can be removed.
2. chk.reachMaxSize()

File: executor/executor.go [hunk 5]
```diff
@@ -571,6 +594,67 @@ func (e *SelectionExec) Next(goCtx goctx.Context) (Row, error) {
...
+			return nil
+		}
+		e.selected, err = expression.VectorizedFilter(e.ctx, e.filters, e.childrenResults[0], e.selected)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+}
+
+func (e *SelectionExec) unBatchedNextChunk(chk *chunk.Chunk) error {
```
Comment (XuHuaiyu) [inline executor/executor.go]: add a comment for this func.

**Comment by XuHuaiyu** [comment]
LGTM
/run-all-tests

**Comment by tiancaiamao** [comment]
LGTM

**Comment by zz-jason** [comment]
to #5261

**Pull request merged**